mrp_production_byproduct_cost_share/models/product.py

```diff
@@ -0,0 +1,77 @@
+# Copyright 2023 ForgeFlow S.L. (https://www.forgeflow.com)
+# License LGPL-3.0 or later (https://www.gnu.org/licenses/lgpl-3.0).
+
+from odoo import models
+from odoo.tools import float_round
+
+
+class ProductTemplate(models.Model):
+    _inherit = "product.template"
+
+    def _compute_bom_count(self):
+        for product in self:
+            super()._compute_bom_count()
+            product.bom_count += self.env["mrp.bom"].search_count(
+                [("byproduct_ids.product_id.product_tmpl_id", "=", product.id)]
+            )
+
+
+class ProductProduct(models.Model):
+    _inherit = "product.product"
+
+    def _compute_bom_count(self):
+        for product in self:
+            super()._compute_bom_count()
+            product.bom_count += self.env["mrp.bom"].search_count(
+                [("byproduct_ids.product_id", "=", product.id)]
+            )
+
+    def _set_price_from_bom(self, boms_to_recompute=False):
+        super()._set_price_from_bom(boms_to_recompute=False)
+        byproduct_bom = self.env["mrp.bom"].search(
+            [("byproduct_ids.product_id", "=", self.id)],
+            order="sequence, product_id, id",
+            limit=1,
+        )
+        if byproduct_bom:
+            price = self._compute_bom_price(
+                byproduct_bom, boms_to_recompute=boms_to_recompute, byproduct_bom=True
+            )
+            if price:
+                self.standard_price = price
+
+    def _compute_bom_price(self, bom, boms_to_recompute=False, byproduct_bom=False):
+        price = super()._compute_bom_price(bom, boms_to_recompute)
+        if byproduct_bom:
+            byproduct_lines = bom.byproduct_ids.filtered(
+                lambda b: b.product_id == self and b.cost_share != 0
+            )
+            product_uom_qty = 0
+            for line in byproduct_lines:
+                product_uom_qty += line.product_uom_id._compute_quantity(
+                    line.product_qty, self.uom_id, round=False
+                )
+            byproduct_cost_share = sum(byproduct_lines.mapped("cost_share"))
+            if byproduct_cost_share and product_uom_qty:
+                return price * byproduct_cost_share / 100
+        else:
+            byproduct_cost_share = sum(bom.byproduct_ids.mapped("cost_share"))
+            if byproduct_cost_share:
+                price *= float_round(
+                    1 - byproduct_cost_share / 100, precision_rounding=0.0001
+                )
```

Comment:
Payed attention, that `total` variable is not passed from parent method `_compute_bom_price` and instead of using `total`, you are using `price` as result of parent method, that returns `bom.product_uom_id._compute_price(total / bom.product_qty, self.uom_id)`

On line 56 you assume that price is always total / product_uom_qty, which is not always true.

On line 60 I am not sure if it is correct replacing `total` to `price`

Reply:
Hi, i am not very sure what do you mean with 'that total variable is not passed from parent method _compute_bom_price '.
The _compute_price function only does:
    
    def _compute_price(self, price, to_unit):
        self.ensure_one()
        if not self or not price or not to_unit or self == to_unit:
            return price
        if self.category_id.id != to_unit.category_id.id:
            return price
        amount = price * self.factor
        if to_unit:
            amount = amount / to_unit.factor
        return amount

And it returns the unit price(that's why I put the result of the super _compute_bom_price function in the variable price). Same as _compute_bom_price function it returns 0 or 'bom.product_uom_id._compute_price(total / bom.product_qty, self.uom_id)'.
The % to apply to it's the same, as I understand, and it doesn't affect if i am working with totals or price.
Also looking here: https://github.com/odoo/odoo/commit/fd52760266ede3b9353936da8dc23e795f5a5031#diff-01e78aa26a6e2155284a52b0e24f0d1f00d68cfe0ed319828b43f37ce882f723R97, I think what I did is consistent. If not, can you suggest how owuld you do it?